ruby_on_rails/testing/model_testing.md

```diff
@@ -0,0 +1,136 @@
+
+## Creating a factory
+
+Below is an example of a user factory. Notice that the factory will be a User class based off of the name of your factory. You can also define as many attributes as you want and it can be useful to focus on adding only the attributes your testing and adding them as need be.
+
+~~~
+
+# This will guess the User class
+FactoryBot.define do
+  factory :user do
+    first_name { "John" }
+    last_name  { "Doe" }
+    admin { false }
+  end
+end
+
+~~~
+
+When creating a factory bot record you have the choice of using the `create(:user)` which returns a saved instance of the object from your database or `build(:user)` which returns an unsaved instance of an object. You might be wondering how we avoid objects hanging around in the database. Thankfully RSpec includes the ability to clear the database in between tests. However if you do find that RSpec's database clearing isn't fitting your needs here you can check the additional resources for alternatives.
+
+## Example model test
+
+When testing our models we usually want to focus on testing the business logic. Essentially what goes on behind each model that enables the user experience to work correctly. This could mean ensuring promotion tickets prices are correct, only correct form data is accepted, custom user verification etc. Below are a few examples along with an explanation to explain it more in depth. It's important to note that each model should have it's own file in the `spec/models` file. This is helpful for keeping your model tests organized and readable.
+
+This is the failing example where we expect the pass to not pass.
+
+~~~
+
+#first we would want to check our spec/factories file.
+#To make sure it's set up properly.
+#It should look like the following.
+
+FactoryBot.define do
+
+  factory :user do 
+    user_name { "userwithout a number" }
+  end 
+
+end
+
+RSpec.describe User, type: :model do
+  #define test expectations
+  it "verifies the user has a number."
+  #We then build a user.
+    user = build(:user) 
+    #Then we check for the expected outcome.
+    expect(user.save). to eql(false)
+  end
+end
```

Comment:
I don't think we need to create another Factory example here. I believe this should build on the example earlier in the lesson. I am also still lost on testing a user that has a number in the user_name. I'd prefer to test a simple validation that people can easily follow.

I believe these example tests are missing contexts. https://www.betterspecs.org/#contexts

I think context will change up the comments about defining the test expectations & then building user.  Not that we have to use this pattern, but what I find to be helpful is to state the context, build models in that context, state the expected result, then test the expectation. So it would be something like this:

```
RSpec.describe User, type: :model do
  describe '#valid?' do
    context "when the user does not have an email address" do
      user =  build(:user)

      it "will not be valid" do
        expect(user.valid?).to be false
      end
    end
    
    context "when the user has an email address" do
      user =  build(:user, email_address: "jdoe@example.com")

      it "will be valid" do
        expect(user.valid?).to be true
      end
    end
  end
end
```

Reply:
Maybe we should include the code that is requiring the number in the username, requiring a email address, or other validation. This might be the missing piece that makes it hard for me to follow along with the example.

Reply:
@rlmoser99 if i'm understanding correctly. For example, a password that requires a special character. Showing the code in the app where that's required, followed by the model code itself, and then the test code itself as the final step. Something like this was what you had in mind?

Reply:
Yes, something like that. I don't think we needs to show the entire model, just the line that the test will be testing. And using the same Factory that we created earlier in the lesson.